### PR TITLE
Add log1mexp.

### PR DESCRIPTION
**Pull request by carlosgmartin**
Adds log1mexp, which computes $\log(1 - \exp(-x))$ in a numerically stable way.

Also handles the gradient correctly.

References:
- https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf
- https://www.tensorflow.org/probability/api_docs/python/tfp/math/log1mexp
- https://www.pymc.io/projects/docs/en/stable/api/generated/pymc.math.log1mexp.html
- https://www.rdocumentation.org/packages/VGAM/versions/1.1-13/topics/log1mexp
- https://juliastats.org/LogExpFunctions.jl/stable/#LogExpFunctions.log1mexp
- https://stdlib.io/docs/api/latest/@stdlib/math/iter/special/log1mexp

Context: https://github.com/jax-ml/jax/pull/29343#issuecomment-3098593780

### CODE REVIEW

File: jax/_src/nn/functions.py [hunk 1]
```diff
@@ -1447,3 +1447,21 @@ def scaled_dot_general(
...
 
   return out
+
+def log1mexp(x: ArrayLike) -> Array:
+  r"""Numerically stable calculation of :math:`\log(1 - \exp(-x))`.
+
+  Based on `TensorFlow's implementation <https://www.tensorflow.org/probability/api_docs/python/tfp/math/log1mexp>`_.
+
+  References:
+    .. [1] Martin Mächler. `Accurately Computing log(1 − exp(−|a|)) Assessed by the Rmpfr package.
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:1491]: From this reference, it looks like the method below is only valid when `x` is non-negative. What happens here when `x` is negative? Will it compute the right thing?

Comment (carlosgmartin) [reply jax/_src/nn/functions.py:1491]: Like log, the function is undefined for negative values. I added a note about that to the docstring, as well as a test that checks that negative values yield NaN.

File: docs/ffi/librms_norm.dylib
Comment (jakevdp) [inline docs/ffi/librms_norm.dylib:1]: Can we remove this binary file?

**Comment by jakevdp** [comment]
Thanks, looks pretty good.

The one question I have here is where we should expose this function. Just to enumerate the possibilities:

- `jax.lax.special` may be an option, but typically these APIs lower directly to single primitives, which `log1mexp` does not
- `jax.numpy` or `jax.scipy.special` may be good options, but typically these would be limited to APIs provided by `numpy` or `scipy.special`, and neither defines `log1mexp`
- `jax.nn` is probably the third best option, but it's somewhat confusing because `log1mexp` is not an activation function. But maybe we can consider it a NN-related utility

All told, I think `jax.nn` is probably the best place for now.

**Review by jakevdp — APPROVED** [review]

File: tests/nn_test.py [hunk 2]
```diff
@@ -720,6 +720,19 @@ def f(hx, _):
     with jax.checking_leaks():
       fwd()  # doesn't crash
 
+  def testLog1mExp(self):
+    self.assertEqual(nn.log1mexp(0), -jnp.inf)
+    self.assertEqual(nn.log1mexp(1), -0.45867515)
```
Comment (jakevdp) [inline tests/nn_test.py]: We should avoid exact equality assertions for floating point values, due to expected differences in rounding error across backends and platforms.

It would be more clear to put all inputs/expected outputs into a single array, and use a single `self.assertAllClose` call.

File: tests/nn_test.py [hunk 3]
```diff
@@ -720,6 +720,19 @@ def f(hx, _):
...
+    self.assertAllClose(nn.log1mexp(1e2), 0.0)
+    self.assertAllClose(nn.log1mexp(1e-5), jnp.log(1e-5))
+    self.assertTrue(jnp.isnan(nn.log1mexp(-1)))
+    self.assertTrue(jnp.isnan(nn.log1mexp(-1e-2)))
+    self.assertTrue(jnp.isnan(nn.log1mexp(-1e2)))
+
+  @parameterized.parameters(0.0, 1e-5, 1.0, 1e-2, 1e2, 1e5)
+  def testLog1mExpGrad(self, x):
+    self.assertAllClose(jax.grad(nn.log1mexp)(x), jnp.reciprocal(jnp.expm1(x)))
```
Comment (jakevdp) [inline tests/nn_test.py]: Similar here: rather than parameterizing over scalars (which generates multiple tests) we could use a single test over an array containing all these scalar values (with `vmap(grad(f))`).

File: jax/_src/nn/functions.py [hunk 4]
```diff
@@ -1447,3 +1447,23 @@ def scaled_dot_general(
   )
 
   return out
+
+def log1mexp(x: ArrayLike) -> Array:
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:1483]: last question here: would it be useful to define a `custom_jvp` rule for this? I'm not sure that tracing through the `cond` is helpful in terms of the gradient, which is always just `1 / expm1(x)`. What do you think?

File: jax/_src/nn/functions.py [hunk 5]
```diff
@@ -1447,3 +1447,23 @@ def scaled_dot_general(
...
+    .. [1] Martin Mächler. `Accurately Computing log(1 − exp(−|a|)) Assessed by the Rmpfr package.
+      <https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf>`_.
+  """
+  numpy_util.check_arraylike("log1mexp", x)
+  x = jnp.asarray(x)
+  c = jnp.log(2.0)
+  return jnp.where(
+      x < c,
+      jnp.log(-jnp.expm1(-x.clip(max=c))),
+      jnp.log1p(-jnp.exp(-x.clip(min=c))),
```
Comment (pearu) [inline jax/_src/nn/functions.py]: Why is clip used here? If `x < c` then `x.clip(max=c)` is effectively a no-op as nothing is clipped under the `x < c` branch, for instance.

Comment (jakevdp) [reply jax/_src/nn/functions.py]: I suspect it's to avoid `NaN` in autodiff (i.e. this issue: https://docs.jax.dev/en/latest/faq.html#gradients-contain-nan-where-using-where)

Comment (pearu) [reply jax/_src/nn/functions.py]: Thanks for the link!

So, to be autodiff-friendly, all expressions in `where` should evaluate to a non-nan value within a function domain.

In the given case, the domain of `log1mexp` is `x > 0` and `log(1 - exp(-x))` is always finite for all positive values of x. The same holds for `log(-expm1(-x))` and `log1p(-exp(-x))`. Neither of the `where` branches (`x < c` and `x >= c`) do not produce non-finite values for all valid inputs. What I am missing?

Is the `clip` introduced to handle a single test case
```
self.assertEqual(nn.log1mexp(0), -jnp.inf)
```
? If so, then the two clip calls can be optimized to a single clip:
```python
r = jnp.where(
      x < c,
      jnp.log(-jnp.expm1(-x)),
      jnp.log1p(-jnp.exp(-x)))
return jnp.where(x == 0, -jnp.inf, r)
```
(warnings: untested and defining log1mexp at 0 is an extension to `x >= 0` domain).

Comment (pearu) [reply jax/_src/nn/functions.py]: Update: I now see that my suggestion above may not be safe.

**Comment by carlosgmartin** [comment]
@jakevdp Regarding [your comment](https://github.com/jax-ml/jax/pull/30811#issuecomment-3164347363) about placement, I had the exactly same thoughts, so I put it under `jax.nn`.

File: jax/_src/nn/functions.py (hunk 5 above)
Comment (carlosgmartin) [reply jax/_src/nn/functions.py]: Looks like the first clip can be removed, but not the second (doing so causes a `nan` at `0` under `grad`).

Comment (pearu) [reply jax/_src/nn/functions.py]: That makes sense. Thanks for checking this out!

File: jax/_src/nn/functions.py (hunk 4 above)
Comment (carlosgmartin) [reply jax/_src/nn/functions.py:1483]: Makes sense. I added a custom jvp.

File: jax/_src/nn/functions.py (hunk 5 above)
Comment (carlosgmartin) [reply jax/_src/nn/functions.py]: I [added a custom jvp](https://github.com/jax-ml/jax/pull/30811#discussion_r2263068567), so it looks like clipping is no longer necessary.

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @carlosgmartin!

File: tests/nn_test.py [hunk 6]
```diff
@@ -720,6 +720,28 @@ def f(hx, _):
...
+        [1e-5, jnp.log(1e-5)],
+        [-1, jnp.nan],
+        [-1e-2, jnp.nan],
+        [-1e2, jnp.nan],
+        [jnp.inf, 0.0],
+    ]).T
+    got = nn.log1mexp(x)
+    self.assertAllClose(got, expected)
+
+  def testLog1mExpGrad(self):
```
Comment (jakevdp) [inline tests/nn_test.py:755]: Now that we're using `custom_jvp`, this test is not as useful (it's basically just an assertion of an identity)

For functions that use `custom_jvp`, we should probably test them with `jtu.check_grads` to check the logic of the custom autodiff rule.

File: jax/_src/scipy/stats/gumbel_r.py [hunk 7]
```diff
@@ -265,5 +255,5 @@ def logsf(x: ArrayLike, loc: ArrayLike = 0, scale: ArrayLike = 1) -> Array:
   neg_z = lax.div(lax.sub(loc, x), scale)
   t1 = lax.neg(lax.exp(neg_z))
   # numerical stability
-  log_sf = _log1mexp(t1)
+  log_sf = log1mexp(-t1)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/gumbel_r.py]: nit: remove the `-` here as well as the `lax.neg` above (or possibly just inline the whole thing: the `t1` variable and the `# numerical stability` comment are not all that helpful)

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Tests of the new function are failing.

**Comment by carlosgmartin** [comment]
Looks like a tolerance issue. What `atol` and `rtol` would you recommend for `testLog1mExp` and `testLog1mExpGrad`?

**Comment by jakevdp** [comment]
> Looks like a tolerance issue. What `atol` and `rtol` would you recommend for `testLog1mExp` and `testLog1mExpGrad`?

That's hard to say – how did you compute the golden values used in the test? How close to those values should we expect JAX to get?

**Comment by carlosgmartin** [comment]
Ok, I'll go with `rtol = atol = 1e-3` for now. Hopefully that works on all hardware.

File: tests/nn_test.py [hunk 8]
```diff
@@ -736,6 +736,31 @@ def testLogMeanExp(self, shape, use_where, keepdims):
...
+        [1, -0.45867515],
+        [1e2, 0.0],
+        [1e-5, jnp.log(1e-5)],
+        [-1, jnp.nan],
+        [-1e-2, jnp.nan],
+        [-1e2, jnp.nan],
+        [jnp.inf, 0.0],
+    ]).T
+    got = nn.log1mexp(x)
+    self.assertAllClose(got, expected, rtol=1e-3, atol=1e-3)
```
Comment (jakevdp) [inline tests/nn_test.py:753]: `atol=1E-3` will dwarf the `1E-5` test case. If that case is important, you should use a smaller absolute tolerance. If it's not important, you should remove it.

Comment (jakevdp) [reply tests/nn_test.py:753]: Oh, wait it's the log of `1E-5` – this is probably OK.

**Review by jakevdp — APPROVED** [review]
